sdks/python/apache_beam/pvalue.py

```diff
@@ -308,7 +308,7 @@ def __getitem__(self, tag):
         self.producer.add_output(pcoll, tag)
     else:
       # Main output is output of inner ParDo.
-      pcoll = self.producer.parts[0].outputs[None]
+      pcoll = self.producer.parts[0].outputs[None]  # type: PCollection
```

Comment:
This doesn't seem safe to me. We should not be promoting the type from `PValue` to a `PCollection` unless we can actually guarantee that it will always be a `PCollection` here.  As far as I can tell, it may actually be `PValue`, but correct me if I'm wrong.

Here's how I came to that conclusion.

The code where `AppliedPTransform.outputs` is assigned values seems to indicate it should be a mapping of `PValue`:

```python
  def add_output(self,
                 output,  # type: Union[pvalue.DoOutputsTuple, pvalue.PValue]
                 tag=None  # type: Union[str, int, None]
                ):
    # type: (...) -> None
    if isinstance(output, pvalue.DoOutputsTuple):
      self.add_output(output[output._main_tag])
    elif isinstance(output, pvalue.PValue):
      # TODO(BEAM-1833): Require tags when calling this method.
      if tag is None and None in self.outputs:
        tag = len(self.outputs)
      assert tag not in self.outputs
      self.outputs[tag] = output
    elif isinstance(output, dict):
      for output_tag, out in output.items():
        self.add_output(out, tag=output_tag)
    else:
      raise TypeError("Unexpected output type: %s" % output)
```

As a result, I have typed `AppliedPTransform.outputs` as `Dict[Union[str, int, None], pvalue.PValue]`.  If it's the case that `AppliedPTransform.outputs` only holds `PCollections` then we should enforce that in the `isinstance` check above (and elsewhere).  

---

Also note that the trailing `# type: PCollection` used here has no effect in mypy, and in fact it generates an error.  In mypy, you can only use this the first time that a variable is defined (e.g. for `pcoll` that's line 300 above this), and you can only use it to broaden a type (as I did originally, going from `PCollection` to `PValue`).   Narrowing the type is basically disregarding/overriding the inspected type, so that requires using `typing.cast()`.

This is why I'm really wary of supporting pytype.

Reply:
If everyone is in agreement that `AppliedPTransform.outputs` can only hold `PCollections` then the typing for `AppliedPTransform` should be fixed.  Doing it this way -- simply overriding the type -- is a superficial fix that leaves the door open for this problem to happen elsewhere in the code where `AppliedPTransform` is used  (and as I mentioned, it doesn't actually work for mypy).

Reply:
IIUC, DoOutputsTuple could only follow a ParDo and its outputs should be PCollection's. However this contract is not really captured in the code. You are right that AppliedPTransform.outputs could have types of PValue, but this should not be the case for DoOutputsTuple.

Reply:
understood.  If we're confident about that contract I think the proper way to solve this would be something like this:

```python
      pval = self.producer.parts[0].outputs[None]
      assert isinstance(pval, PCollection), "DoOutputsTuple should follow a ParDo"
      pcoll = pval
```
 
If you want to be a bit more formal you could raise a `TypeError` or the like.  

That will naturally narrow the type of `pcoll`.  The reason for the temporary variable `pval` is that you cannot assign of type `PValue` to a variable already defined as `PCollection` (above in the `if tag is not None` block).

Reply:
The types here are correct, `self._pcolls` should be a map to PCollections. This PR looks like an improvement to me. Would you be OK with this going in (with the assertion to do the type restriction)? 

(FWIW, it'd be nice to clean up the whole DoOutputsTuple thing, and how inputs and outputs are represented in general, but that's another issue.)

Reply:
yep, looks good to me.